lib.rs

```diff
@@ -24,6 +25,23 @@ mod token {
                 data: PSP22Data::new(supply, Self::env().caller()),
             }
         }
+
+        fn emit_events(&self, events: ink::prelude::vec::Vec<PSP22Event>) {
+            for event in events {
+                match event {
+                    PSP22Event::Transfer(from, to, value) => {
+                        self.env().emit_event(Transfer { from, to, value })
```

Comment:
not sure if `Transfer` type is available outside contract module, but if so, you might want to implement conversion (`Into`) from `PSP22Event` type to this 'native' event type

the advantage is that we encapsulate event type matching close to the event definition, and outside, like here, we can just go `for e in events { self.env().emit_event(e.into()) }`

but this might be infeasible (due to opaque event types) and is pretty much personal taste, so feel free to ignore this note comment :)

Reply:
Unfortunately events are not visible outside the scope of `ink::contract` macro... yet. AFAIK ink! 5 will introduce a possibility to define events externally. After that happens, the whole enum->event conversion will be removed and `PSP22Data` methods will simply return vectors of native events. Before that happens, automating the conversion would require putting some very cryptic boilerplate inside the contract. Since this is meant to be simple educational example, I'd rather go with stupid&simple approach.